genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/stages/DetermineJobFinalStatusStage.java

```diff
@@ -0,0 +1,103 @@
+/*
+ *
+ *  Copyright 2020 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.agent.execution.statemachine.stages;
+
+import com.netflix.genie.agent.execution.process.JobProcessResult;
+import com.netflix.genie.agent.execution.statemachine.ExecutionContext;
+import com.netflix.genie.agent.execution.statemachine.ExecutionStage;
+import com.netflix.genie.agent.execution.statemachine.FatalTransitionException;
+import com.netflix.genie.agent.execution.statemachine.RetryableTransitionException;
+import com.netflix.genie.agent.execution.statemachine.States;
+import com.netflix.genie.common.dto.JobStatusMessages;
+import com.netflix.genie.common.external.dtos.v4.JobStatus;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.EnumSet;
+
+/**
+ * Sends the server the final job status, if the job reached a state where it is appropriate to do so.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class DetermineJobFinalStatusStage extends ExecutionStage {
+
+    private static final EnumSet<JobStatus> INTERMEDIATE_ACTIVE_STATUSES = EnumSet.of(
```

Comment:
Don't we already have this via the `getActiveStatuses()` static function in `JobStatus`? Seems like we wouldn't want to split this kind of logic around multiple places?

Reply:
This is what we throw when the server response doesn't make sense (which includes a connection error). e.g.:
https://github.com/Netflix/genie/blob/0a90dc76858f14d7b43fd1b92f68758412126244/genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentJobServiceImpl.java#L314

If the server clearly responds with "This spec failed to resolve", or "Job already claimed, etc" then no point in retrying.
But if we get a GenieRuntimeException, then it's likely this is a transient failure and should be retried.